File: code/queue/redis_queue.py
class RQueue(object):
    """ redis queue wrap """
    KEY = "redis"
    def __init__(self, redis_conn):
        """
        :host: redis host
        :port: redis port
        :redis_conn: redis connection
        """
        self.redis = redis_conn
        #redis.Redis(host=host, port=port, db=db)

    def get(self, timeout=5):
        """Get element from q."""
        value = self.redis.blpop(self.KEY, timeout=timeout)
        return value and value[1] or value

    def put(self, value):
        """Put element into q."""
        self.redis.rpush(self.KEY, value)

    def empty(self):
        """Return True if q is empty."""
        return False if self.redis.llen(self.KEY) else True
        
    def count(self):
        """Return the number of emelments in the q."""
        return self.redis.llen(self.KEY)

    def __str__(self):
        return "Redis Queue: len {}".format(self.redis.llen(self.KEY))

    def __repr__(self):
        return "{0}('{host}', '{port}', '{db}')".format(
                              self.__class__.__name__,
                              **self.redis.connection_pool.connection_kwargs)

File: code/queue/redis_queue.rb
# encoding: utf-8

require 'redis'
require 'redis/connection/hiredis'

class RedisQueue
  attr_reader :redis, :queue, :buffer, :timeout
  attr_accessor :buffer_max

  def initialize(queue_name, options = {})
    if !queue_name.is_a?(String) || queue_name.empty?
      raise ArgumentError, "Queue name must be a non empty string"
    end

    @queue      = "#{queue_name}:source"
    @buffer     = "#{queue_name}:buffer"
    @redis      = options.fetch(:redis, Redis.current)
    @buffer_max = options.fetch(:conc_num, 8)
  end

  # Return the buffer max num
  #
  # @return [int]
  alias :conc_num :buffer_max

  # Reset the buffer max num
  #
  # @return [int]
  alias :conc_num= :buffer_max=

  # Returns the number of queue elements
  #
  # @return [int]
  def length
    redis.llen(queue)
  end
  alias :size :length

  # Checks whether the queue is empty
  #
  # @return [true, false]
  def empty?
    !(length > 0)
  end
  alias :empty :empty?

  # Access the first element of queue
  #
  # @return [Object]
  def front
    redis.lindex(queue, 0)
  end

  # Access the last element of queue
  #
  # @return [Object]
  def back
    redis.lindex(queue, -1)
  end

  # Access all elements of queue
  #
  # @return [List of Object]
  def list
    redis.lrange(queue, 0, -1)
  end

  # Access all elements of buffer
  #
  # @return [List of Object]
  def runners
    redis.lrange(buffer, 0, -1)
  end

  # Removes the first element of queue and push it to buffer
  # if the buffer is fill, return nil
  #
  # @return [Object]
  def pop
    redis.rpoplpush(queue, buffer) unless buffer_is_fill?
  end

  # Inserts element at the end of queue and return the size of queue
  #
  # @return [int]
  def push(obj)
    redis.lpush(queue, obj)
  end
  alias :<< :push

  # Remove one element of queue
  # return false if the element not exist
  #
  # @return [true, false]
  def remove(obj)
    (redis.lrem(queue, 1, obj) != 0) ? true : false
  end
  alias :delete :remove

  # Remove one element of buffer
  # return false if the element not exist
  #
  # @return [true, false]
  def commit(obj)
    (redis.lrem(buffer, 1, obj) != 0) ? true : false
  end

  # Clear the queue and buffer
  def clear
    redis.del(queue)
    redis.del(buffer)
  end

  private

  def buffer_is_fill?
    redis.llen(buffer) >= buffer_max
  end
end
